src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1954,201 @@ MutableBigInteger sqrt() {
+        MutableBigInteger dividend = sr[1];
+        dividend.shiftAddDisjoint(getBlockZimmermann(1, len, limit, blockLen), blockLen);
+        MutableBigInteger twiceSqrt = new MutableBigInteger(sr[0]);
+        twiceSqrt.leftShift(1);
+        MutableBigInteger q = new MutableBigInteger();
+        MutableBigInteger u = dividend.divide(twiceSqrt, q);
+
+        MutableBigInteger sqrt = sr[0];
+        sqrt.shiftAdd(q, blockLen);
+
+        MutableBigInteger chunk = u; // Corresponds to ub + a_0 in the paper
+        chunk.shiftAddDisjoint(getBlockZimmermann(0, len, limit, blockLen), blockLen);
+        BigInteger qBig = q.toBigInteger(); // Cast to BigInteger to use fast multiplication
+        MutableBigInteger qSqr = new MutableBigInteger(qBig.multiply(qBig).mag);
+
+        MutableBigInteger rem;
+        if (needRemainder) {
+            rem = chunk;
+            if (rem.subtract(qSqr) == -1) {
+                twiceSqrt = new MutableBigInteger(sqrt);
+                twiceSqrt.leftShift(1);
+
+                rem.add(ONE);
+                rem.subtract(twiceSqrt);
+                sqrt.subtract(ONE);
+            }
+        } else {
+            rem = null;
+            if (chunk.compare(qSqr) == -1)
+                sqrt.subtract(ONE);
+        }
+
+        // Unnormalize
+        if (excessInts != 0) {
+            final int halfShift = excessInts << 4;
+            if (needRemainder && !rem.isZero()) {
+                final int s0Ints = (halfShift + 31) >> 5;
+                MutableBigInteger s0 = sqrt.getBlockZimmermann(0, sqrt.intLen, sqrt.offset + sqrt.intLen, s0Ints);
+                if (s0.intLen == s0Ints)
+                    s0.value[0] &= -1 >>> -halfShift; // Remove excess bits
+
+                if (!s0.isZero()) { // MutableBigInteger.multiply() assumes intLen != 0
+                    MutableBigInteger twiceS0 = new MutableBigInteger(s0);
+                    twiceS0.leftShift(1);
+                    MutableBigInteger doubleProd = new MutableBigInteger();
+                    twiceS0.multiply(sqrt, doubleProd); // Use classic multiplication
+
+                    MutableBigInteger s0Sqr = new MutableBigInteger();
+                    s0.multiply(s0, s0Sqr);
+
+                    rem.add(doubleProd);
+                    rem.subtract(s0Sqr);
+                }
+                rem.intLen -= excessInts;
+            }
+            sqrt.rightShift(halfShift);
+        }
+        return new MutableBigInteger[] { sqrt, rem };
+    }
+
+    /**
+     * Returns a {@code MutableBigInteger} obtained taking {@code blockLen} ints from
+     * {@code this} number, starting at {@code blockIndex*blockLen}.<br/>
+     * Used by Zimmermann square root.
+     * @param blockIndex the block index
+     * @param len the logical length of the input value in units of 32 bits
+     * @param limit the offset which is the end of valid words in the input value
+     * @param blockLen the length of the block in units of 32 bits
+     */
```

Comment:
At first, I couldn't understand that `blockIndex` goes in the opposite direction as the array indices. Also, "starting" is very confusing, because no `int` at `blockIndex*blockLen` gets copied AFAIU.

Two improvements:
* Add a `@return` tag. Alternatively, use comments other than JavaDoc `/**`.
* Make it explicit that the direction of `blockIndex` is the opposite of the array indices, and reconsider the wording like "starting", etc.

Reply:
For the sake of information: the documentation of this method is actually based on that of  `MBI.getBlock(int, int, int)`, which has the same problems.

Reply:
I agree, and one could keep improving MBI's doc a lot, in particular implicit assumptions.
But let's concentrate on this PR and make this code readable and well documented.